fc/code_generation.py

```diff
@@ -59,6 +60,67 @@ def load_template(*name):
     return env.get_template(path)
 
 
+class DataInterpolation(Quantity):
+    """Represents linear interpolation on data loaded from file, for code generation."""
+    _next_id = 0  # Used to generate unique table IDs in code
+
+    # Sympy annoyingly overwrites __new__
+    def __new__(cls, name, data, index_variable, *args, **kwargs):
+        print('interp__new__', name)
+        obj = super().__new__(cls, name, real=True)
+
+        # Record a unique ID for this table
+        obj._id = cls._next_id
+        cls._next_id += 1
+
+        # Ensure we depend on the index variable in the model's graph
+        obj._args = [index_variable]
+
+        return obj
+
+    def __init__(self, name, data, index_variable, units):
+        """Create a new interpolation construct.
+
+        :param name: the data file base name
+        :param data: 2d numpy array containing the column-wise data
+        :param index_variable: the :class:`cellmlmanip.model.Variable` used to index the data
+        :param units: the units of the interpolated values (a :class:`~cellmlmanip.units.UnitStore.Unit`)
+        """
+        print('interp__init__', name, index_variable, units)
+        super().__init__(name, units)
+        self.data = data[1]
+        self.index_variable = index_variable
+        self.initial_index = '%.17g' % data[0, 0]
```

Comment:
What's going on here? String representations of the first and final data points?

Reply:
Yep, for insertion into generated code.